polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -28,305 +31,365 @@ struct ParallelLICM : public ParallelLICMBase<ParallelLICM> {
+       memoryEffects.freesMemory()) &&
+      hasConflictsInLoop(op, loop)) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "cannot be hoisted: found conflicting operation\n");
     return false;
+  }
 
   // Recurse into the regions for this op and check whether the contained ops
-  // can be hoisted.
-  // We can inductively assume that this op will have its block args available
-  // outside the loop
+  // can be hoisted. We can inductively assume that this op will have its block
+  // args available outside the loop
   SmallPtrSet<Operation *, 2> willBeMoved2(willBeMoved.begin(),
                                            willBeMoved.end());
-  willBeMoved2.insert(op);
-  /*
-  for (auto &region : op->getRegions())
-    for (auto &block : region)
-       for (auto arg : block.getArguments())
-           willBeMoved2.insert(&arg);
-           */
-
-  for (auto &region : op->getRegions()) {
-    for (auto &block : region) {
-      for (auto &innerOp : block)
-        if (!canBeParallelHoisted(&innerOp, scope, willBeMoved2)) {
-          return false;
-        } else
-          willBeMoved2.insert(&innerOp);
+  willBeMoved2.insert(&op);
+
+  for (Region &region : op.getRegions()) {
+    for (Operation &innerOp : region.getOps()) {
+      if (!canBeHoisted(innerOp, loop, willBeMoved2))
+        return false;
+      willBeMoved2.insert(&innerOp);
     }
   }
-  return true;
-}
 
-void moveParallelLoopInvariantCode(scf::ParallelOp looplike) {
-  auto &loopBody = looplike.getLoopBody();
+  LLVM_DEBUG(llvm::dbgs().indent(2) << "can be hoisted: no conflicts found\n");
 
-  // We use two collections here as we need to preserve the order for insertion
-  // and this is easiest.
-  SmallPtrSet<Operation *, 8> willBeMovedSet;
-  SmallVector<Operation *, 8> opsToMove;
+  return true;
+}
 
+// Populate \p opsToMove with operations that can be hoisted out of the given
+// loop \p loop.
+template <typename T, typename = std::enable_if_t<llvm::is_one_of<
+                          T, scf::ParallelOp, AffineParallelOp>::value>>
+static void collectHoistableOperation(T loop,
+                                      SmallVectorImpl<Operation *> &opsToMove) {
   // Do not use walk here, as we do not want to go into nested regions and hoist
   // operations from there. These regions might have semantics unknown to this
   // rewriting. If the nested regions are loops, they will have been processed.
-  for (auto &block : loopBody) {
-    for (auto &op : block.without_terminator()) {
-      if (canBeParallelHoisted(&op, looplike, willBeMovedSet)) {
-        opsToMove.push_back(&op);
-        willBeMovedSet.insert(&op);
-      }
+  SmallPtrSet<Operation *, 8> willBeMoved;
+  for (Block &block : loop.getLoopBody()) {
+    for (Operation &op : block.without_terminator()) {
+      if (!canBeHoisted(op, loop, willBeMoved))
+        continue;
+      opsToMove.push_back(&op);
+      willBeMoved.insert(&op);
     }
```

Comment:
Can we use walkers here?

Reply:
no the original comment indicates we do not want to walk nested regions.